src/Server.cpp

```diff
@@ -4,193 +4,559 @@
+ * is in the connections map.
+ *
+ * @param event The epoll event to handle.
+ */
+void handleEvent(Server& server, struct epoll_event event)
+{
+	uint32_t eventMask = event.events;
+
+	std::map<int, Socket>::const_iterator iter = server.getVirtualServers().find(event.data.fd);
+	if (iter != server.getVirtualServers().end()) {
+		if ((eventMask & EPOLLERR) != 0) {
+			LOG_ERROR << "Error condition happened on the associated file descriptor of " << iter->second;
+			close(event.data.fd);
+			return;
 		}
+		acceptConnections(server, iter->first, iter->second, eventMask);
+	} else {
+		if ((eventMask & EPOLLERR) != 0) {
+			LOG_DEBUG << "epoll_wait: EPOLLERR";
+			eventMask = EPOLLIN;
+		} else if ((eventMask & EPOLLHUP) != 0) {
+			LOG_DEBUG << "epoll_wait: EPOLLHUP";
+			eventMask = EPOLLIN;
+		}
+		handleConnection(server, event.data.fd, server.getConnections().at(event.data.fd));
 	}
 }
 
 /**
- * @brief Handle data from a client connection.
+ * @brief Accept new connections
+ *
+ * If the event mask does not contain `EPOLLIN`, it logs an error and returns. A server socket should always receive a
+ * read event.
  *
- * This method is responsible for reading data from a client socket
- * and processing it.
+ * In a loop, calls Server::accept() to accept new connections on the server socket.
+ * When Server::accept() is successful, information of the client socket is retrieved with Server::retrieveSocketInfo().
+ * This new client socket is then registered with Server::registerConnection().
+ *
+ * The while-loop allows for multiple connections to be accepted in one call of this function. It is broken when accept
+ * sets errno to EAGAIN or EWOULDBLOCK (equivalent error codes indicating that a non-blocking operation would normally
+ * block), meaning that no more connections are pending. If another errno is returned logs an error and continues to the
+ * next iteration of the loop. This is because the server sockets are in EPOLLET mode. If only one connection would be
+ * accepted the same server fd would not be reported again.
+ *
+ * @param server The server object to accept connections for.
+ * @param serverFd File descriptor of the server socket.
+ * @param serverSock Server socket which reported an event.
+ * @param eventMask Event mask of the reported event.
+ */
+void acceptConnections(Server& server, int serverFd, const Socket& serverSock, uint32_t eventMask)
+{
+	LOG_DEBUG << "Accept connections on: " << serverSock;
+
+	if ((eventMask & EPOLLIN) == 0) {
+		LOG_ERROR << "Received unknown event:" << eventMask;
+		return;
+	}
+
+	while (true) {
+		struct sockaddr_storage clientAddr = {};
+		socklen_t clientLen = sizeof(clientAddr);
+
+		// NOLINTNEXTLINE: we need to use reinterpret_cast to convert sockaddr_storage to sockaddr
+		struct sockaddr* addrCast = reinterpret_cast<struct sockaddr*>(&clientAddr);
+
+		const int clientFd = server.accept(serverFd, addrCast, &clientLen);
```

Comment:
I called the wrong function:
SocketPolicy has a member function which just calls c-function accept(2). It is also called accept (maybe not that smart). Since I called it as just accept(...) it lead to a recursive call of SocketPolicy::accept, leading to segfault.

I fixed it by calling the correct one as `::accept` (global namespace).

Alternative: name the function something else. Before a refactor I had it as acceptConnection(), but this got confusing since Server already has acceptConnection**s**().

Since the correct function is called again, it should work

Reply:
True. Looked into the function but did not figure out, that it leads to a recursive call. Now that's obvious. 

Works as intended now. 

Perhaps you could rename it to "acceptSingleConnection"?

Reply:
Makes sense: renamed to acceptSingleConnection